### PR TITLE
Add a GCS mirror to WORKSPACE URLs, along with a script to update them.

### PR DESCRIPTION
**Pull request by rmmh**
**What this PR does / why we need it**:
We get occasional flakes when WORKSPACE downloads fail. This should make that much more unlikely, by having two independent endpoints to try.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: hack/update-bazel-gcs-mirror.sh
Comment (ixdy) [inline hack/update-bazel-gcs-mirror.sh]: s/cache/mirror/?

File: build/workspace_mirror.bzl
Comment (rmmh) [reply build/workspace_mirror.bzl]: Yes!

File: hack/update-bazel-gcs-mirror.sh
Comment (fejta) [inline hack/update-bazel-gcs-mirror.sh]: Can we fold this into a `bazel run //:update-workspace-cache` rule?

Given that this is bazel-centric, providing a bazel-centric UX seems appropriate.

**Review by fejta — COMMENTED** [review]
This is pretty sweet
/lgtm
/hold

File: hack/update-bazel-gcs-mirror.sh
Comment (rmmh) [reply hack/update-bazel-gcs-mirror.sh]: Woops, missed this when renaming things. Fixed!

File: hack/update-bazel-gcs-mirror.sh
Comment (rmmh) [reply hack/update-bazel-gcs-mirror.sh]: Would need to be unsandboxed for gsutil to work, but I guess it could have a rule?

File: hack/update-workspace-mirror.sh
Comment (ixdy) [inline hack/update-workspace-mirror.sh]: this one is a little unfortunate. :\

File: hack/update-bazel-gcs-mirror.sh
Comment (ixdy) [reply hack/update-bazel-gcs-mirror.sh]: `bazel run --direct_run` might be closer to what you want?

File: hack/update-workspace-mirror.sh
Comment (rmmh) [reply hack/update-workspace-mirror.sh]: the alternative hacks are probably worse :/

**Comment by rmmh** [comment]
/hold cancel

PTAL

File: hack/update-workspace-mirror.sh
Comment (fejta) [reply hack/update-workspace-mirror.sh]: Let bazel process the file for you:
`WORKSPACE`
```
load("//somewhere:something.bzl", "foo")
foo()
```
`bash$ bazel run //:update_http_archive_cache`
`somewhere/something.bzl`
```
def foo():
  urls = []
  for k, v in native.existing_rules().items():
    if v["kind"] in ["http_archive", "new_http_archive"]:
      print("MATCH", k, "url", v.get("url"), "urls", v.get("urls"))
      u = v.get("url")
      us = v.get("urls")
      if u:
        urls.append(u)
      if us:
        urls.extend(us)
    else:
      # print("no", k)
      pass
  if urls:
    print("lets update", urls)
  fail("demo over!")
```

https://docs.bazel.build/versions/master/skylark/macros.html
https://docs.bazel.build/versions/master/skylark/lib/native.html#existing_rules

Comment (fejta) [reply hack/update-workspace-mirror.sh]: Move `foo()` into a `bar/BUILD` and change `foo()` to generate a binary rule:

`bar/BUILD.bazel`
```
load("somewhere/something.bzl", "foo")
foo("update-cache")
```

`something.bzl`
```
def foo(name):
  ...
  if urls:
    print("lets update", urls)
    native.sh_binary(
      name = name,
      srcs = ["cache.sh"],  # the curl | gsutil line in this file
      args = urls,
    )
```

Then `bazel run //bar:update-cache`

https://docs.bazel.build/versions/master/skylark/cookbook.html#macro_native
https://docs.bazel.build/versions/master/be/shell.html#sh_binary

Comment (fejta) [reply hack/update-workspace-mirror.sh]: Create a [genrule](https://docs.bazel.build/versions/master/be/general.html#genrule) that takes as input the list of URLs we use and outputs the list of URLs that don't have a cached version. 

 
`foo.bzl`
```
def cached_url(u):
  return PREFIX + u

def foofunc():
  urls = ... # see above
  native.genrule(
    name="blah",
    outs=["missing-cache.urls"],
    cmd="""
urls=( {urls} )
touch $@
for u in urls; do
  curl --fail --head $u || echo $u >> $@
done
  """.format(urls=" ".join([cached_url(u) for u in urls]))
```


`bar/BUILD`
```
load(":foo.bzl", "foofunc")
foofunc() # generates the blah rule that generates the file
```

`bazel build //bar:blah && cat bazel-genfiles/bar/missing-cache.urls`

Comment (fejta) [reply hack/update-workspace-mirror.sh]: Make an `sh_test()` rule which ensure that either a) there are no uncached urls or b) it can successfully call the update_cache rule to update them

```
load("//somewhere:something.bzl", "foo")
foofunc()
foo("update-cache")

sh_test(
  name = "verify_cache_test",
  srcs = ["verify_cache_test.sh"], # [[ -z "$(cat $1)" ]] && exit 0 || $2 <<<$1 
  args = ["$(location :missing-cache.urls)", "$(location :update-cache)"],
  local = True, # use user gsutil creds
)
```

https://docs.bazel.build/versions/master/be/shell.html#sh_test

Comment (fejta) [reply hack/update-workspace-mirror.sh]: If we do all of this then whenever someone changes the WORKSPACE file presubmits tests will fail unless the author (or someone else with write access to the cache, hopefully not the presubmit service account) has run the update_cache binary with the new url list.

All of this (or something like it) would be most useful across time, but all is not necessary for this PR.

Can we at least do something so bazel is the one processing the urls rather than `grep` and `sed`?

**Comment by rmmh** [comment]
Implemented a much cleaner bazel-based update: `bazel run //build:update-mirror`

File: build/workspace_mirror.bzl
Comment (fejta) [reply build/workspace_mirror.bzl]: I think this will print on every bazel command

File: build/workspace_mirror.bzl
Comment (fejta) [inline build/workspace_mirror.bzl]: May want to explore whether this is necessary?

File: build/workspace_mirror.bzl
Comment (fejta) [inline build/workspace_mirror.bzl]: Could we set the default value here to mirror_urls() and eliminate the macro?

File: hack/update-workspace-mirror.sh
Comment (fejta) [inline hack/update-workspace-mirror.sh]: Does this still exit non zero on failures?

File: build/workspace_mirror.bzl
Comment (fejta) [inline build/workspace_mirror.bzl]: Nit use "?

**Review by fejta — COMMENTED** [review]
Lgtm aside from print cleanup. Also some suggestions.

File: build/workspace_mirror.bzl
Comment (rmmh) [reply build/workspace_mirror.bzl]: done

File: build/workspace_mirror.bzl
Comment (rmmh) [reply build/workspace_mirror.bzl]: local is necessary to execute when the dependency graph changes-- there's no other way to say "depends on WORKSPACE"

File: build/workspace_mirror.bzl
Comment (rmmh) [reply build/workspace_mirror.bzl]: done

File: hack/update-workspace-mirror.sh
Comment (rmmh) [reply hack/update-workspace-mirror.sh]: yes-- it just makes sure the tmpfile doesn't linger if someone ctrl-Cs

File: build/workspace_mirror.bzl
Comment (rmmh) [reply build/workspace_mirror.bzl]: Maybe, but evaluating it at load time is a little gross-- it makes the value depending on when it's imported. I also prefer having the target name in WORKSPACE.

**Review by fejta — COMMENTED** [review]
Fantastic!
/lgtm
/hold

File: hack/update-workspace-mirror.sh
Comment (rmmh) [reply hack/update-workspace-mirror.sh]: Piping directly can create empty or partial files if curl fails.

File: build/root/WORKSPACE
Comment (rmmh) [reply build/root/WORKSPACE]: I'd need to implement custom http_archive, new_http_archive, and http_file. I prefer the explicit form with mirror-- it makes it more obvious that something is happening behind the scenes.

**Review by BenTheElder — APPROVED** [review]
/lgtm
this is awesome, we should consider using this elsewhere

File: hack/update-workspace-mirror.sh
Comment (BenTheElder) [inline hack/update-workspace-mirror.sh]: we should automate running this regularly once this is in

**Comment by ixdy** [comment]
ah, I found a different bug: the `urls.txt` is missing a trailing newline, so the `while read` bash loop wasn't breaking out because it couldn't read the golang url; it just wasn't reading the last URL (https://storage.googleapis.com/kubernetes-release/network-plugins/cni-plugins-amd64-v0.6.0.tgz).

**Comment by rmmh** [comment]
If it fails to download a URL, it continues without failing, so that go URL not mirroring is fine.

**Pull request merged**
